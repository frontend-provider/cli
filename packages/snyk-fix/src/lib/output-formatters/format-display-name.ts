import * as pathLib from 'path';
import { Identity } from '../../types';

export function formatDisplayName(
  path: string,
  identity: Pick<Identity, 'type' | 'targetFile'>,
): string {
  if (!identity.targetFile) {
    return `${identity.type} project`;
  }
  // show paths relative to where `snyk fix` is running
  return pathLib.relative(
    process.cwd(),
    pathLib.join(path, identity.targetFile),
  );
}
