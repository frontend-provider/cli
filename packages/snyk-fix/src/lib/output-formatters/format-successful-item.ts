import * as chalk from 'chalk';

import { EntityToFix, FixChangesSummary } from '../../types';
import { PADDING_SPACE } from './show-results-summary';

/*
 * Generate formatted output that describes what changes were applied, which failed.
 */
export function formatChangesSummary(
  entity: EntityToFix,
  changes: FixChangesSummary[],
): string {
  return `${PADDING_SPACE}${
    entity.scanResult.identity.targetFile
  }\n${changes.map((c) => formatAppliedChange(c)).join('\n')}`;
}

// TODO:
// write test for these
function formatAppliedChange(change: FixChangesSummary): string | null {
  if (change.success === true) {
    return `${PADDING_SPACE}${chalk.green('âœ”')} ${change.userMessage}`;
  }
  if (change.success === false) {
    return `${PADDING_SPACE}${chalk.red('x')} ${chalk.red(
      change.userMessage,
    )}\n${PADDING_SPACE}Reason:${PADDING_SPACE}${change.reason}${
      change.tip ? `.\n${PADDING_SPACE}Tip:     ${change.tip}` : undefined
    }`;
  }
  return '';
}
