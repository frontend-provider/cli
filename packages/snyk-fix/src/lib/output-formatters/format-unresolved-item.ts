import * as chalk from 'chalk';

import { EntityToFix } from '../../types';
import { formatDisplayName } from './format-display-name';
import { PADDING_SPACE } from './show-results-summary';

export function formatUnresolved(
  entity: EntityToFix,
  userMessage: string,
  tip?: string,
): string {
  const name = formatDisplayName(
    entity.workspace.path,
    entity.scanResult.identity,
  );
  const tipMessage = tip ? `\n${PADDING_SPACE}Tip:     ${tip}` : '';
  const errorMessage = `${PADDING_SPACE}${name}\n${PADDING_SPACE}${chalk.red(
    'âœ–',
  )} ${chalk.red(userMessage)}`;
  return errorMessage + tipMessage;
}
