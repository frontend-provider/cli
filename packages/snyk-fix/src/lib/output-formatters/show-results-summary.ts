import * as chalk from 'chalk';

import { FixHandlerResultByPlugin } from '../../plugins/types';
import { ErrorsByEcoSystem } from '../../types';
import { convertErrorToUserMessage } from '../errors/error-to-user-message';
import { formatChangesSummary } from './format-successful-item';
import { formatUnresolved } from './format-unresolved-item';
export const PADDING_SPACE = '  '; // 2 spaces

export async function showResultsSummary(
  resultsByPlugin: FixHandlerResultByPlugin,
  exceptionsByScanType: ErrorsByEcoSystem,
): Promise<string> {
  const successfulFixesSummary = generateSuccessfulFixesSummary(
    resultsByPlugin,
  );
  const unresolvedSummary = generateUnresolvedSummary(
    resultsByPlugin,
    exceptionsByScanType,
  );
  const overallSummary = generateFixedAndFailedSummary(
    resultsByPlugin,
    exceptionsByScanType,
  );
  return `${successfulFixesSummary}\n\n${unresolvedSummary}\n\n${overallSummary}`;
}

export function generateSuccessfulFixesSummary(
  resultsByPlugin: FixHandlerResultByPlugin,
): string {
  const sectionTitle = 'Successful fixes:';
  const formattedTitleHeader = `${chalk.bold(sectionTitle)}`;
  let summary = '';

  for (const plugin of Object.keys(resultsByPlugin)) {
    const fixedSuccessfully = resultsByPlugin[plugin].succeeded;
    if (fixedSuccessfully.length > 0) {
      summary +=
        '\n\n' +
        fixedSuccessfully
          .map((s) => formatChangesSummary(s.original, s.changes))
          .join('\n\n');
    }
  }
  if (summary) {
    return formattedTitleHeader + summary;
  }
  return chalk.red('âœ– No successful fixes');
}

export function generateUnresolvedSummary(
  resultsByPlugin: FixHandlerResultByPlugin,
  exceptionsByScanType: ErrorsByEcoSystem,
): string {
  const sectionTitle = 'Unresolved items:';
  const formattedTitleHeader = `${chalk.bold(sectionTitle)}`;
  let summary = '';

  for (const plugin of Object.keys(resultsByPlugin)) {
    const skipped = resultsByPlugin[plugin].skipped;
    if (skipped.length > 0) {
      summary +=
        '\n\n' +
        skipped
          .map((s) => formatUnresolved(s.original, s.userMessage))
          .join('\n\n');
    }
    const failed = resultsByPlugin[plugin].failed;
    if (failed.length > 0) {
      summary +=
        '\n\n' +
        failed
          .map((s) =>
            formatUnresolved(s.original, convertErrorToUserMessage(s.error)),
          )
          .join('\n\n');
    }
  }

  if (Object.keys(exceptionsByScanType).length) {
    for (const ecosystem of Object.keys(exceptionsByScanType)) {
      const unresolved = exceptionsByScanType[ecosystem];
      summary +=
        '\n\n' +
        unresolved.originals
          .map((s) => formatUnresolved(s, unresolved.userMessage))
          .join('\n\n');
    }
  }
  if (summary) {
    return formattedTitleHeader + summary;
  }
  return '';
}

export function generateFixedAndFailedSummary(
  resultsByPlugin: FixHandlerResultByPlugin,
  exceptionsByScanType: ErrorsByEcoSystem,
): string {
  const sectionTitle = 'Summary:';
  const formattedTitleHeader = `${chalk.bold(sectionTitle)}`;
  let fixedItems = 0;
  let failedItems = 0;
  for (const plugin of Object.keys(resultsByPlugin)) {
    fixedItems += resultsByPlugin[plugin].succeeded.length;
  }

  for (const plugin of Object.keys(resultsByPlugin)) {
    const results = resultsByPlugin[plugin];
    failedItems += results.failed.length + results.skipped.length;
  }

  if (Object.keys(exceptionsByScanType).length) {
    for (const ecosystem of Object.keys(exceptionsByScanType)) {
      const unresolved = exceptionsByScanType[ecosystem];
      failedItems += unresolved.originals.length;
    }
  }

  return `${formattedTitleHeader}\n\n${PADDING_SPACE}${chalk.bold.red(
    failedItems,
  )} items were not fixed\n${PADDING_SPACE}${chalk.green.bold(
    fixedItems,
  )} items were successfully fixed`;
}
