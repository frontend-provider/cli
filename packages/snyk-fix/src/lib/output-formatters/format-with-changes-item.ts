import * as chalk from 'chalk';

import { EntityToFix, FixChangesSummary } from '../../types';
import { formatDisplayName } from './format-display-name';
import { PADDING_SPACE } from './show-results-summary';

/*
 * Generate formatted output that describes what changes were applied, which failed.
 */
export function formatChangesSummary(
  entity: EntityToFix,
  changes: FixChangesSummary[],
): string {
  return `${PADDING_SPACE}${formatDisplayName(
    entity.workspace.path,
    entity.scanResult.identity,
  )}\n${changes.map((c) => formatAppliedChange(c)).join('\n')}`;
}

function formatAppliedChange(change: FixChangesSummary): string | null {
  if (change.success === true) {
    return `${PADDING_SPACE}${chalk.green('âœ”')} ${change.userMessage}`;
  }
  if (change.success === false) {
    return `${PADDING_SPACE}${chalk.red('x')} ${chalk.red(
      change.userMessage,
    )}\n${PADDING_SPACE}Reason:${PADDING_SPACE}${change.reason}${
      change.tip ? `.\n${PADDING_SPACE}Tip:     ${change.tip}` : undefined
    }`;
  }
  return '';
}
