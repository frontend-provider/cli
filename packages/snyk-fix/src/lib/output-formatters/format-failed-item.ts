import { FailedToFix, isWithError } from '../../plugins/types';

import { convertErrorToUserMessage } from '../errors/error-to-user-message';
import { formatChangesSummary } from './format-with-changes-item';
import { formatUnresolved } from './format-unresolved-item';

export function formatFailed(failed: FailedToFix): string {
  if (isWithError(failed)) {
    return formatUnresolved(
      failed.original,
      convertErrorToUserMessage(failed.error),
      failed.tip,
    );
  }
  return formatChangesSummary(failed.original, failed.changes);
}
