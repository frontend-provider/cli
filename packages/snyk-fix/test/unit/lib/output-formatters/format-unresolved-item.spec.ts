import stripAnsi = require('strip-ansi');
import { formatUnresolved } from '../../../../src/lib/output-formatters/format-unresolved-item';
import { generateEntityToFix } from '../../../helpers/generate-entity-to-fix';

describe('format unresolved item', () => {
  it('formats unresolved as expected by default', () => {
    const entity = generateEntityToFix(
      'pip',
      'requirements.txt',
      JSON.stringify({}),
    );
    const res = formatUnresolved(entity, 'Failed to process item');
    expect(stripAnsi(res)).toMatchSnapshot();
  });

  it('formats ok when missing targetFile', () => {
    const entity = generateEntityToFix(
      'npm',
      undefined as any,
      JSON.stringify({}),
    );
    const res = formatUnresolved(entity, 'Failed to process item');
    expect(stripAnsi(res)).toMatchSnapshot();
  });

  it('formats ok with tip', () => {
    const entity = generateEntityToFix('pip', 'Pipfile', JSON.stringify({}));
    const res = formatUnresolved(
      entity,
      'Failed to fix',
      'Make sure you have pipenv installed',
    );
    expect(stripAnsi(res)).toMatchSnapshot();
  });
});
