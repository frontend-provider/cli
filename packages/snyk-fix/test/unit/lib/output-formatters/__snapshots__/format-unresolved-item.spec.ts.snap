// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format unresolved item formats ok when missing targetFile 1`] = `
"  npm project
  ✖ Failed to process item"
`;

exports[`format unresolved item formats ok with tip 1`] = `
"  Pipfile
  ✖ Failed to fix
  Tip:     Make sure you have pipenv installed"
`;

exports[`format unresolved item formats unresolved as expected by default 1`] = `
"  requirements.txt
  ✖ Failed to process item"
`;
